go/worker/beacon/worker.go

```diff
@@ -0,0 +1,588 @@
+		case beacon.StateReveal:
+			w.onStateReveal(ev, scrapeState)
+		case beacon.StateComplete:
+			w.onStateComplete(ev, scrapeState)
+		}
+	}
+}
+
+func (w *Worker) eventOkForState(ev *beacon.SCRAPEEvent, expectedState beacon.RoundState) bool {
+	if w.state == nil {
+		return false
+	}
+	if ev.Epoch != w.state.Epoch || ev.Round != w.state.Round {
+		w.logger.Error("epoch/round mismatch",
+			"epoch", w.state.Epoch,
+			"round", w.state.Round,
+		)
+		return false
+	}
+	if w.state.State != expectedState {
+		w.logger.Error("round state mismatch",
+			"state", w.state.State,
+		)
+		return false
+	}
+
+	return true
+}
+
+func (w *Worker) onStateCommit(ev *beacon.SCRAPEEvent, scrapeState *beacon.SCRAPEState) {
+	if w.state != nil {
+		if ev.Epoch == w.state.Epoch && ev.Round == w.state.Round {
+			// This shouldn't happen, log something about it.
+			w.logger.Error("commit: received redundant round event, ignoring")
+			return
+		}
+		w.persistState(nil)
+	}
+
+	// New round.
+
+	// Determine if we are present in the participant list, and ensure
+	// that the point for each participant are valid.
+	var (
+		isPresent bool
+		points    []scrape.Point
+	)
+	for _, id := range ev.Participants {
+		if id.Equal(w.identity.NodeSigner.Public()) {
+			isPresent = true
+		}
+		node, err := w.registry.GetNode(
+			w.ctx,
+			&registry.IDQuery{
+				Height: ev.Height,
+				ID:     id,
+			},
+		)
+		if err != nil {
+			w.logger.Error("commit: failed to query node descriptor",
+				"err", err,
+				"id", id,
+			)
+			return
+		}
+		if node.Beacon == nil {
+			w.logger.Error("commit: participant node missing point",
+				"id", id,
+			)
+			return
+		}
+		points = append(points, node.Beacon.Point)
+	}
+	h, cmpH := hash.NewFrom(points), hash.NewFrom(scrapeState.Instance.Participants)
+	if !h.Equal(&cmpH) {
+		w.logger.Error("commit: participant point list mismatch")
+		return
+	}
+	if !isPresent {
+		w.logger.Info("commit: node is not present in participant list")
+		return
+	}
+
+	// Initialize the local SCRAPE instance.
+	instance, err := scrape.New(&scrape.Config{
+		PrivateKey:    &w.identity.BeaconScalar,
+		Participants:  points,
+		Threshold:     scrapeState.Instance.Threshold,
+		PVSSThreshold: scrapeState.Instance.PVSSThreshold,
+	})
+	if err != nil {
+		w.logger.Error("commit: failed to initialize SCRAPE instance",
+			"err", err,
+		)
+		return
+	}
+
+	// Generate the commit.
+	commit, err := instance.Commit()
+	if err != nil {
+		w.logger.Error("commit: failed to generate commit",
+			"err", err,
+		)
+		return
+	}
+	commitPayload := beacon.SCRAPECommit{
+		Epoch:  ev.Epoch,
+		Round:  ev.Round,
+		Commit: commit,
+	}
+	tx := transaction.NewTransaction(0, nil, beacon.MethodSCRAPECommit, commitPayload)
+
+	// Persist so it is possible to recover.
+	w.persistState(&roundState{
+		Epoch:    ev.Epoch,
+		Round:    ev.Round,
+		State:    ev.State,
+		Instance: instance,
+		CommitTx: tx,
+	})
+
+	// Submit the commit tx.
+	// XXX: Add a retry?
```

Comment:
will especially need this if we go with the design where you have to wait to be proposer before you can submit your commit/reveal

Reply:
I mean, this is in the issue todo list for a reason.